tools/health-monitor/monitor/file_cache/cache_monitor.go

```diff
@@ -0,0 +1,311 @@
+				if err != nil {
+					log.Err("cache_monitor::cacheWatcher : [%v]", err)
+					return
+				}
+
+				if strings.ToUpper(event.Op.String()) == create {
+					fc.createEvent(&event)
+
+				} else if strings.ToUpper(event.Op.String()) == remove {
+					fc.removeEvent(&event)
+
+				} else if strings.ToUpper(event.Op.String()) == chmod {
+					fc.chmodEvent(&event)
+
+				} else if strings.ToUpper(event.Op.String()) == write {
+					fc.writeEvent(&event)
+
+				} else if strings.ToUpper(event.Op.String()) == rename {
+					fc.renameEvent(&event)
+
+				} else if strings.ToUpper(event.Op.String()) == move {
+					fc.moveEvent(&event)
+
+				}
+
+			case err := <-w.Error:
+				log.Err("cache_monitor::cache_watcher : [%v]", err)
+				return
+
+			case <-w.Closed:
+				return
+			}
+		}
+	}()
+
+	// watch file cache directory for changes
+	if err := w.Add(fc.tmpPath); err != nil {
+		log.Err("cache_monitor::cacheWatcher : [%v]", err)
+		return err
+	}
+
+	// set recursive watcher on file cache directory
+	if err := w.AddRecursive(fc.tmpPath); err != nil {
+		log.Err("cache_monitor::cacheWatcher : [%v]", err)
+		return err
+	}
+
+	// list of all of the files and folders currently being watched
+	for path := range w.WatchedFiles() {
+		log.Debug("cache_monitor::cacheWatcher : Watching %v", path)
+	}
+
+	// Start the watching process - it'll check for changes every 100ms
+	if err := w.Start(time.Millisecond * 100); err != nil {
+		log.Err("cache_monitor::cacheWatcher : [%v]", err)
+		return err
+	}
+
+	return nil
+}
+
+func (fc *FileCache) createEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileRemovedMap, event.Path)
+		fc.cacheObj.fileCreatedMap[event.Path] = event.Size()
+		fc.cacheObj.cacheSize += event.Size()
+		fc.cacheObj.cacheConsumed = (float64)(fc.cacheObj.cacheSize*100) / (fc.maxSizeMB * common.MbToBytes)
+	}
+
+	e := fc.getCacheEventObj(event)
+	if !event.IsDir() {
+		e.Value[fileSize] = strconv.FormatInt(event.Size(), 10)
+	}
+
+	log.Debug("cache_monitor::createEvent : %v", e)
+	fc.ExportStats(time.Now().Format(time.RFC3339), e)
+}
+
+func (fc *FileCache) removeEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileCreatedMap, event.Path)
+		fc.cacheObj.fileRemovedMap[event.Path] = event.Size()
+		fc.cacheObj.cacheSize = int64(math.Max(0, float64(fc.cacheObj.cacheSize-event.Size())))
+		fc.cacheObj.cacheConsumed = (float64)(fc.cacheObj.cacheSize*100) / (fc.maxSizeMB * common.MbToBytes)
+	}
+
+	e := fc.getCacheEventObj(event)
+	if !event.IsDir() {
+		e.Value[fileSize] = strconv.FormatInt(event.Size(), 10)
+	}
+
+	log.Debug("cache_monitor::removeEvent : %v", e)
+	fc.ExportStats(time.Now().Format(time.RFC3339), e)
+}
+
+func (fc *FileCache) chmodEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileRemovedMap, event.Path)
+		fileSize := fc.cacheObj.fileCreatedMap[event.Path]
+
+		if fileSize == event.Size() {
+			return
```

Comment:
does this mean we skip lines 232+ if the file size didnt change with the mode? I don't understand why we do that.

Reply:
same for the writeEvent method below

Reply:
Nice observation. I corrected the part for chmod events where we will export those events.
For write events, I am not logging and exporting it because there can be lot of write events for large files. On the other hand, blobfuse is send the upload and download progress to the monitor binary for every 100MB of data transferred.